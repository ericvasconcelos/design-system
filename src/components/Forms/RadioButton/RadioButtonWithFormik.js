import React from 'react'

import { Text } from 'components/Text'
import { Field, ErrorMessage } from 'formik'
import { bool, func, string, arrayOf, shape } from 'prop-types'

import { Label, StyledInput } from './styles'

export const RadioButtonWithFormik = ({ name, label, options, onChange, disabled, className }) => (
  <>
    <Text tag="p" styleType="body2" styleColor="greyCash400">
      {label}
    </Text>
    <Field name={name} className={className}>
      {({ field }) =>
        options.map(option => (
          <Label key={option.label} htmlFor={`option-${option.value}`}>
            <StyledInput
              id={`option-${option.value}`}
              name={name}
              disabled={disabled}
              type="radio"
              value={option.value}
              onChange={e => {
                onChange(e)
                field.onChange(e)
              }}
              onBlur={field.onBlur}
              defaultChecked={field.value === option.value}
            />
            <Text tag="span" styleType="body2" styleColor="greyCash400">
              {option.label}
            </Text>
          </Label>
        ))}
    </Field>
    <ErrorMessage name={name}>
      {messageError => (
        <Text tag="p" styleType="body3" styleColor="error400">
          {messageError}
        </Text>
      )}
    </ErrorMessage>
  </>
)

RadioButtonWithFormik.propTypes = {
  name: string.isRequired,
  label: string.isRequired,
  options: arrayOf(
    shape({
      label: string.isRequired,
      value: string.isRequired
    })
  ).isRequired,
  onChange: func,
  disabled: bool,
  className: string
}

RadioButtonWithFormik.defaultProps = {
  disabled: false,
  onChange: () => {},
  className: ''
}
