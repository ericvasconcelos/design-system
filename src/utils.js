export { arrayCombinations } from 'utils/arrayCombinations'
export { getBackgroundColor } from 'utils/getBackgroundColor'
export { getBorder } from 'utils/getBorder'
export { getButtonStyle } from 'utils/getButtonStyle'
export { getBorderRadius } from 'utils/getBorderRadius'
export { getColorFromName } from 'utils/getColorFromName'
export { getGuttersNames } from 'utils/getGuttersNames'
export { getIconNames } from 'utils/getIconNames'
export { getPadding } from 'utils/getPadding'
export { getPaddingNames } from 'utils/getPaddingNames'
export { getTypography } from 'utils/getTypography'
export { parsePalleteName } from 'utils/parsePalleteName'
export { styleColorValidator } from 'utils/styleColorValidator'
export { useAriaProps } from 'utils/useAriaProps'

export { DateUtils } from 'utils/DateUtils'
export { QueryUtils } from 'utils/QueryUtils'
export { ObjectUtils }  from 'utils/ObjectUtils'
export { CookiesUtils } from 'utils/CookiesUtils'
export { MoneyUtils } from 'utils/MoneyUtils'
export { NumberUtils } from 'utils/NumberUtils'