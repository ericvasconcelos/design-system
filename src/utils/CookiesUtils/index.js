export class CookiesUtils {
  static toObject(string = '') {
    const cookies = {}
    if (string !== '') {
      const parts = string.split('; ')
      parts.forEach(part => {
        const [key, value] = part.split('=')
        cookies[key] = value
      })
    }
    return cookies
  }
}
