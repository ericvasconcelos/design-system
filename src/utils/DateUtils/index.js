export class DateUtils {
  static toDate(string, format = 'pt-br') {
    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = string?.split('/')
        break
      case 'iso':
        [year, month, day] = string?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    return new Date(year, parseInt(month, 10) - 1, day)
  }

  static getAge(date) {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const difference = Date.now() - date.getTime()

    return Math.abs(new Date(difference).getUTCFullYear() - 1970)
  }

  static toString(date, format = 'pt-br') {
    if (!(date instanceof Date)) {
      throw new TypeError('Date invalid')
    }

    const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()
    const month = date.getMonth() + 1 < 10 ? `0${date.getMonth() + 1}` : date.getMonth() + 1
    const year = date.getFullYear()

    switch (format) {
      case 'pt-br':
        return `${day}/${month}/${year}`
      case 'iso':
        return `${year}-${month}-${day}`
      default:
        throw new Error('Format invalid')
    }
  }

  static validate(date, format = 'pt-br') {
    const monthDaysMap = {
      '01': 31,
      '02': 29,
      '03': 31,
      '04': 30,
      '05': 31,
      '06': 30,
      '07': 31,
      '08': 31,
      '09': 30,
      10: 31,
      11: 30,
      12: 31
    }
    let isValid = false

    let day = ''
    let month = ''
    let year = ''

    switch (format) {
      case 'pt-br':
        [day, month, year] = date?.split('/')
        break
      case 'iso':
        [year, month, day] = date?.split('-')
        break
      default:
        throw new Error('Format invalid')
    }

    isValid = isValid || parseInt(day, 10) <= monthDaysMap[month]
    isValid = isValid && parseInt(year, 10) > 1900

    return isValid
  }

  static convertMonthToYear(month) {
    return parseInt(month, 10) / 12
  }

  static formatPeriod(period, periodType) {
    const periodTypeMap = {
      monthly: { singular: 'mÃªs', plural: 'meses' },
      yearly: { singular: 'ano', plural: 'anos' },
      daily: { singular: 'dia', plural: 'dias' }
    }
    return `${period} ${periodTypeMap[periodType][period < 2 ? 'singular' : 'plural']}`
  }
}
