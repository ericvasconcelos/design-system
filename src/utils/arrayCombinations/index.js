export const arrayCombinations = (...args) => {
  const returnable = []
  const max = args.length - 1
  function helper(arr, i) {
    for (let j = 0, { length } = args[i]; j < length; j++) {
      const newArray = arr.slice(0) // clone arr
      newArray.push(args[i][j])
      if (i === max) returnable.push(newArray)
      else helper(newArray, i + 1)
    }
  }
  helper([], 0)
  return returnable
}
