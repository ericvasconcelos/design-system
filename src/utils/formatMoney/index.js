import { toMoney, toNumber } from 'vanilla-masker'

export const formatMoney = entry => {
  const moneyFormatter = {
    precision: 2,
    separator: ',',
    delimiter: '.',
    unit: 'R$'
  }

  return toMoney(toNumber(entry), moneyFormatter)
}
