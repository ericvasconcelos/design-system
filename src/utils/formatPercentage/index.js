import { toMoney, toNumber } from 'vanilla-masker'

export const formatPercentage = entry => {
  const percentageFormatter = {
    precision: 1,
    separator: ',',
    suffixUnit: '%'
  }
  return toMoney(toNumber(entry), percentageFormatter)
}
