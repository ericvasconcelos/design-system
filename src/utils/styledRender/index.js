import React from 'react'

import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { theme } from 'styles/theme'

export const styledRender = (children, options = {}) => {
  const rendered = render(<ThemeProvider theme={theme}>{children}</ThemeProvider>, options)
  return {
    ...rendered,
    rerender: (ui, rerenderOptions = {}) =>
      styledRender(ui, { container: rendered.container, ...rerenderOptions })
  }
}
