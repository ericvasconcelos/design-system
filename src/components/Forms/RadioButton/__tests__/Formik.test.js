import React from 'react'

import userEvent from '@testing-library/user-event'
import { Formik, Form } from 'formik'
import { styledRender } from 'utils/styledRender'

import { RadioButtonWithFormik } from '../RadioButtonWithFormik'

describe('should render the component correctly when', () => {
  const options = [
    { label: 'One', value: 'one' },
    { label: 'Two', value: 'two' },
    { label: 'Three', value: 'three' }
  ]

  it('has basic props', () => {
    const { container, getByText, getByLabelText } = styledRender(
      <Formik initialValues={{ example: '' }}>
        <Form>
          <RadioButtonWithFormik name="example" label="Example" options={options} />
        </Form>
      </Formik>
    )

    expect(getByLabelText(/one/i)).not.toBeChecked()
    expect(getByLabelText(/two/i)).not.toBeChecked()
    expect(getByLabelText(/three/i)).not.toBeChecked()
    expect(getByText('Example')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has checked prop', () => {
    const { container, getByLabelText } = styledRender(
      <Formik initialValues={{ example: 'one' }}>
        <Form>
          <RadioButtonWithFormik name="example" label="Example" options={options} />
        </Form>
      </Formik>
    )

    expect(getByLabelText(/one/i)).toBeChecked()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has disabled prop', () => {
    const onChange = jest.fn()
    const { container, getByLabelText } = styledRender(
      <Formik initialValues={{ example: '' }}>
        <Form>
          <RadioButtonWithFormik
            name="example"
            label="Example"
            options={options}
            disabled
            onChange={onChange}
          />
        </Form>
      </Formik>
    )

    userEvent.click(getByLabelText(/one/i))

    expect(onChange).not.toBeCalled()
    expect(getByLabelText(/one/i)).toHaveAttribute('disabled')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('has error prop', () => {
    const { container, getByText } = styledRender(
      <Formik
        initialValues={{ example: '' }}
        initialErrors={{ example: 'Erro' }}
        initialTouched={{ example: true }}
      >
        <Form>
          <RadioButtonWithFormik name="example" label="Example" options={options} />
        </Form>
      </Formik>
    )

    expect(getByText('Erro')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('click call onChange prop', () => {
    const onChange = jest.fn()
    const { getByLabelText } = styledRender(
      <Formik initialValues={{ example: '' }}>
        <Form>
          <RadioButtonWithFormik
            name="example"
            label="Example"
            options={options}
            onChange={onChange}
          />
        </Form>
      </Formik>
    )

    userEvent.click(getByLabelText(/one/i))

    expect(getByLabelText(/one/i)).toBeChecked()
    expect(onChange).toBeCalled()
  })
})
