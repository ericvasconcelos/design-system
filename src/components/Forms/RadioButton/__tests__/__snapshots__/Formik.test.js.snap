// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component correctly when has basic props 1`] = `
.c0 {
  color: #69788C;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 50%;
  display: inline-block;
  position: relative;
  padding: 6px;
  margin: 0.25rem 12px 0.25rem 0;
}

.c2:not([disabled]):hover,
.c2:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c2[checked] {
  padding: 3px;
  border-width: 5px;
  border-color: #01B36A;
  background-color: #FFFFFF;
}

<form
  action="#"
>
  <p
    class="c0"
    data-testid="Text"
  >
    Example
  </p>
  <label
    class="c1"
    for="option-one"
  >
    <input
      class="c2"
      id="option-one"
      name="example"
      type="radio"
      value="one"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      One
    </span>
  </label>
  <label
    class="c1"
    for="option-two"
  >
    <input
      class="c2"
      id="option-two"
      name="example"
      type="radio"
      value="two"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Two
    </span>
  </label>
  <label
    class="c1"
    for="option-three"
  >
    <input
      class="c2"
      id="option-three"
      name="example"
      type="radio"
      value="three"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Three
    </span>
  </label>
</form>
`;

exports[`should render the component correctly when has checked prop 1`] = `
.c0 {
  color: #69788C;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 50%;
  display: inline-block;
  position: relative;
  padding: 6px;
  margin: 0.25rem 12px 0.25rem 0;
}

.c2:not([disabled]):hover,
.c2:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c2[checked] {
  padding: 3px;
  border-width: 5px;
  border-color: #01B36A;
  background-color: #FFFFFF;
}

<form
  action="#"
>
  <p
    class="c0"
    data-testid="Text"
  >
    Example
  </p>
  <label
    class="c1"
    for="option-one"
  >
    <input
      checked=""
      class="c2"
      id="option-one"
      name="example"
      type="radio"
      value="one"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      One
    </span>
  </label>
  <label
    class="c1"
    for="option-two"
  >
    <input
      class="c2"
      id="option-two"
      name="example"
      type="radio"
      value="two"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Two
    </span>
  </label>
  <label
    class="c1"
    for="option-three"
  >
    <input
      class="c2"
      id="option-three"
      name="example"
      type="radio"
      value="three"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Three
    </span>
  </label>
</form>
`;

exports[`should render the component correctly when has disabled prop 1`] = `
.c0 {
  color: #69788C;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #00224B14;
  border: 2px solid #00224B14;
  border-radius: 50%;
  display: inline-block;
  position: relative;
  padding: 6px;
  margin: 0.25rem 12px 0.25rem 0;
}

.c2:not([disabled]):hover,
.c2:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c2[checked] {
  padding: 3px;
  border-width: 5px;
  border-color: #00224B14;
  background-color: #69788C52;
}

<form
  action="#"
>
  <p
    class="c0"
    data-testid="Text"
  >
    Example
  </p>
  <label
    class="c1"
    for="option-one"
  >
    <input
      class="c2"
      disabled=""
      id="option-one"
      name="example"
      type="radio"
      value="one"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      One
    </span>
  </label>
  <label
    class="c1"
    for="option-two"
  >
    <input
      class="c2"
      disabled=""
      id="option-two"
      name="example"
      type="radio"
      value="two"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Two
    </span>
  </label>
  <label
    class="c1"
    for="option-three"
  >
    <input
      class="c2"
      disabled=""
      id="option-three"
      name="example"
      type="radio"
      value="three"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Three
    </span>
  </label>
</form>
`;

exports[`should render the component correctly when has error prop 1`] = `
.c0 {
  color: #69788C;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c3 {
  color: #D93E3E;
  font-size: 0.8125rem;
  font-weight: 400;
  line-height: 1.056rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 50%;
  display: inline-block;
  position: relative;
  padding: 6px;
  margin: 0.25rem 12px 0.25rem 0;
}

.c2:not([disabled]):hover,
.c2:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c2[checked] {
  padding: 3px;
  border-width: 5px;
  border-color: #01B36A;
  background-color: #FFFFFF;
}

<form
  action="#"
>
  <p
    class="c0"
    data-testid="Text"
  >
    Example
  </p>
  <label
    class="c1"
    for="option-one"
  >
    <input
      class="c2"
      id="option-one"
      name="example"
      type="radio"
      value="one"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      One
    </span>
  </label>
  <label
    class="c1"
    for="option-two"
  >
    <input
      class="c2"
      id="option-two"
      name="example"
      type="radio"
      value="two"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Two
    </span>
  </label>
  <label
    class="c1"
    for="option-three"
  >
    <input
      class="c2"
      id="option-three"
      name="example"
      type="radio"
      value="three"
    />
    <span
      class="c0"
      data-testid="Text"
    >
      Three
    </span>
  </label>
  <p
    class="c3"
    data-testid="Text"
  >
    Erro
  </p>
</form>
`;
